mrphowl/javascript-track
# JS-121: Express Basics - Parameters, Query Strings and Modularizing Routes

# Related Tickets

[JS-121](https://codeandtheory.atlassian.net/browse/JS-121)

# Description

Solutions to code challenges

Lesson work files

- Group routes into different files and import them into the app
- Using middleware to do additional processing and computations
- Handling errors and rendering error page using templates

Flash cards project

- Importing flat files for data
- Add parameter to route
- Access a single card data using the route parameter
- Using query string to determine which side of the card to show (question or answer)
  - handling invalid and empty query string parameter
- Using pug conditionals to show and hide elements
- Picking a card randomly and redirecting to the card's route

## Setup

1. To install dependencies, run `npm install` in `JS-121_express-basics/5-parameters-query-strings-modular-routes/exercise/flashcards`
2. Start the application with `nodemon`

# Validation

1. Open http://localhost:3000 in a browser.
    The page will redirect to http://localhost:3000/hello

    <img width="366" alt="image" src="https://user-images.githubusercontent.com/8042132/183692071-7bb6cdaf-f2cb-4f75-b7e0-0daed75f2106.png">

2. Enter a name and click **Submit**
    The page will redirect back to http://localhost:3000/ and content will be displayed in the page
    - the heading section will be titled **Flash Cards**
    - then name entered will be displayed in the line **Welcome, _NAME_!**
    - and a **goodbye** button is displayed on the next line before
    - content has a single link **_Begin!_** is on the next line it points to http://localhost:3000/cards
    - the footer shows a line **An app to help you study**

    <img width="294" alt="image" src="https://user-images.githubusercontent.com/8042132/183692297-670e83dc-2681-4f12-858b-c7da8cb2b932.png">

3. Click the link **Begin!**
    The page will redirect to a random card [http://localhost:3000/cards/\<id\>?side=question](http://localhost:3000/cards/\<id\>?side=question)
    We can manually check this by replacing the `<id>` with a number from 1 to 5
    The content will display the following:
    - The question as a sub header
    - The hint as paragraph text
    - Two links, each in their own line:
        - **Answer** goes to the route that shows the answer [http://localhost:3000/cards/\<id\>?side=answer](http://localhost:3000/cards/\<id\>?side=answer)
        - **Next card** picks another card

    <img width="650" alt="image" src="https://user-images.githubusercontent.com/8042132/183692588-a358fd1c-46e0-4a5c-b6d4-b070629361c9.png">

4. Click the link **Answer**
    The page will redirect to the same card but shows the answer instead [http://localhost:3000/cards/\<id\>?side=answer](http://localhost:3000/cards/\<id\>?side=answer)
    The content will show the following:
    - The answer as a sub header
    - Two links, in separate lines:
        - **Question** goes to the route that shows the question and hint [http://localhost:3000/cards/\<id\>?side=question](http://localhost:3000/cards/\<id\>?side=question)
        - **Next card** picks another card

    <img width="433" alt="image" src="https://user-images.githubusercontent.com/8042132/183692846-aac8a9ee-3c12-4816-b32a-b76d2a98d272.png">

# Risks

None
